File: backend/app/schemas/properties.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field

from app.schemas.areas import EPCSummary, PlanningSummary
from app.schemas.common import SourceAttribution


class PropertySale(BaseModel):
    transaction_id: str
    price: int
    transfer_date: date
    property_type: str | None = None
    new_build: bool | None = None
    tenure: str | None = None
    ppd_category: str | None = None


class PropertySearchResult(BaseModel):
    property_key: str
    address: str
    postcode: str | None = None
    latest_price: int
    latest_transfer_date: date
    property_type: str | None = None
    transaction_count: int
    source: SourceAttribution


class PropertySearchResponse(BaseModel):
    query: str
    results: list[PropertySearchResult]
    total: int
    dataset_version: str | None = None
    message: str | None = None


class NearbySalesSummary(BaseModel):
    postcode: str
    count: int
    median_price: int | None = None
    minimum_price: int | None = None
    maximum_price: int | None = None


class PropertyDetail(BaseModel):
    property_key: str
    address: str
    postcode: str | None = None
    property_type: str | None = None
    town_city: str | None = None
    district: str | None = None
    county: str | None = None
    sales: list[PropertySale]
    nearby_sales: NearbySalesSummary | None = None
    planning: PlanningSummary
    epc: EPCSummary
    source: SourceAttribution
    limitations: list[str] = Field(default_factory=list)

File: frontend/types/properties.ts
import type { SourceAttribution } from "./common";
import type { EPCSummary, PlanningSummary } from "./areas";

export interface PropertySearchResult {
  property_key: string;
  address: string;
  postcode: string | null;
  latest_price: number;
  latest_transfer_date: string;
  property_type: string | null;
  transaction_count: number;
  source: SourceAttribution;
}

export interface PropertySearchResponse {
  query: string;
  results: PropertySearchResult[];
  total: number;
  dataset_version: string | null;
  message: string | null;
  suggestions: PropertySearchResult[];
}

export interface PropertyDetail {
  property_key: string;
  address: string;
  postcode: string | null;
  property_type: string | null;
  town_city: string | null;
  district: string | null;
  county: string | null;
  sales: {
    transaction_id: string;
    price: number;
    transfer_date: string;
    property_type: string | null;
    new_build: boolean | null;
    tenure: string | null;
    ppd_category: string | null;
  }[];
  nearby_sales: {
    postcode: string;
    count: number;
    median_price: number | null;
    minimum_price: number | null;
    maximum_price: number | null;
  } | null;
  planning: PlanningSummary;
  epc: EPCSummary;
  source: SourceAttribution;
  limitations: string[];
}
